src/vsr/replica.zig

```diff
@@ -179,6 +179,16 @@ pub fn ReplicaType(
         /// Invariant: replica < node_count
         replica: u8,
 
+        /// Runtime upper-bound number of prepares/requests in the pipeline.
+        /// Does not change after initialization.
+        /// Invariants:
+        /// - pipeline_limit ≥ pipeline_prepare_queue_max
+        /// - pipeline_limit ≤ pipeline_prepare_queue_max + pipeline_request_queue_max
```

Comment:
:thinking: could we perhaps make only `pipeline_request_queue_max` part configurable?

I think our pipeline is actually two different numbers:

* how many prepares to pipeline (this is a protocol-defining parameter, as replica routinely make judgement that for op=N op=N-pipeline *must* be comitted)
* how many requests to pipeline on top of that

and only the second one can very

In the implementation, the two are mixed together, because we subdivide the same memory between:

* prepare cache
* prepare pipeline
* request pipeline

but I don't think it makes sense to leak this all the way to the CLI?


In the alterantive I am imagining, the client specifies the length of requests pipeline, and then we add that to `pipeline_prepare_queue_max` to get the actual size?

Do we even need `pipeline_request_queue_max` as a compile-time const? Could we use runtime-allocated slice for pipeline queue/cache instaed of an array?

Reply:
> In the alterantive I am imagining, the client specifies the length of requests pipeline, and then we add that to pipeline_prepare_queue_max to get the actual size?

:+1: Yeah, that works! Correspondingly, I renamed the flag to `--limit-pipeline-requests`.

> Do we even need pipeline_request_queue_max as a compile-time const?

I think it is still a convenient alias for `constants.clients_max - constants.pipeline_prepare_queue_max` (i.e. the upper-bound request queue size).

> Could we use runtime-allocated slice for pipeline queue/cache instaed of an array?

We _could_, but it would be an extensive enough refactor that it isn't worth it imo. The trouble is that we swap `Replica.pipeline` back-and-forth between two different representations:

```
        /// The pipeline is a queue for a replica which is the primary and in status=normal.
        /// At all other times the pipeline is a cache.
        pipeline: union(enum) {
            /// The primary's pipeline of inflight prepares waiting to commit in FIFO order,
            /// with a tail of pending requests which have not begun to prepare.
            /// This allows us to pipeline without the complexity of out-of-order commits.
            queue: PipelineQueue,
            /// Prepares in the cache may be committed or uncommitted, and may not belong to the
            /// current view.
            cache: PipelineCache,
        },
```

Internally, the `cache` uses a single array, while `queue` splits into two different arrays.